connect/src/context_resolver.rs

```diff
@@ -0,0 +1,368 @@
+use crate::state::context::ContextType;
+use crate::{
+    core::{Error, Session},
+    protocol::{
+        autoplay_context_request::AutoplayContextRequest,
+        player::{Context, TransferState},
+    },
+    state::{context::UpdateContext, ConnectState},
+};
+use std::cmp::PartialEq;
+use std::{
+    collections::{HashMap, VecDeque},
+    fmt::{Display, Formatter},
+    hash::{Hash, Hasher},
+    time::Duration,
+};
+use thiserror::Error as ThisError;
+use tokio::time::Instant;
+
+#[derive(Debug, Clone)]
+enum Resolve {
+    Uri(String),
+    Context(Context),
+}
+
+#[derive(Debug, Clone)]
+pub(super) enum ContextAction {
+    Append,
+    Replace,
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct ResolveContext {
+    resolve: Resolve,
+    fallback: Option<String>,
+    update: UpdateContext,
+    action: ContextAction,
+}
+
+impl ResolveContext {
+    fn append_context(uri: impl Into<String>) -> Self {
+        Self {
+            resolve: Resolve::Uri(uri.into()),
+            fallback: None,
+            update: UpdateContext::Default,
+            action: ContextAction::Append,
+        }
+    }
+
+    pub fn from_uri(
+        uri: impl Into<String>,
+        fallback: impl Into<String>,
+        update: UpdateContext,
+        action: ContextAction,
+    ) -> Self {
+        let fallback_uri = fallback.into();
+        Self {
+            resolve: Resolve::Uri(uri.into()),
+            fallback: (!fallback_uri.is_empty()).then_some(fallback_uri),
+            update,
+            action,
+        }
+    }
+
+    pub fn from_context(context: Context, update: UpdateContext, action: ContextAction) -> Self {
+        Self {
+            resolve: Resolve::Context(context),
+            fallback: None,
+            update,
+            action,
+        }
+    }
+
+    /// the uri which should be used to resolve the context, might not be the context uri
+    fn resolve_uri(&self) -> Option<&str> {
+        // it's important to call this always, or at least for every ResolveContext
+        // otherwise we might not even check if we need to fallback and just use the fallback uri
+        match self.resolve {
+            Resolve::Uri(ref uri) => ConnectState::valid_resolve_uri(uri),
+            Resolve::Context(ref ctx) => ConnectState::get_context_uri_from_context(ctx),
+        }
+        .or(self.fallback.as_deref())
```

Comment:
As it doesn't evaluated any values it shouldn't matter if it's eagerly or not. Unless there isn't any benefit to it, I would just keep it as is.

Reply:
Just thinking that this will probably always evaluate to `None` because of the eager evaluation at compile time, not sure if that's the behavior you intend?

Reply:
But why would it always evaluate to `None`, I think eagerly evaluated just means always/early evaluated? As `self.fallback` is an `Option<String>` it should only provide the value in place. I think the only difference would be the evaluation of `as_deref()`. But this is only used here to convert `Option<String>` into `Option<&str>`.

Just want to understand where you are coming from, as from my perspective it shouldn't make a difference.

Reply:
You are right.